Userland/Libraries/LibCore/SessionManagement.cpp

```diff
@@ -0,0 +1,68 @@
+/*
+ * Copyright (c) 2022, Peter Elliott <pelliott@serenityos.org>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include <LibCore/Directory.h>
+#include <LibCore/ProcessStatisticsReader.h>
+#include <LibCore/SessionManagement.h>
+#include <LibCore/System.h>
+
+namespace Core::SessionManagement {
+
+static ErrorOr<Core::ProcessStatistics const*> get_proc(Core::AllProcessesStatistics const& stats, pid_t pid)
+{
+    for (auto& proc : stats.processes) {
+        if (proc.pid == pid)
+            return &proc;
+    }
+    return Error::from_string_literal("Could not find pid in process statistics.");
+}
+
+ErrorOr<pid_t> root_session_id(Optional<pid_t> force_sid)
+{
+    auto stats = Core::ProcessStatisticsReader::get_all(false);
```

Comment:
I'm confused here, why do we need to parse /proc/all instead of calling `getsid` ?

https://pubs.opengroup.org/onlinepubs/9699919799/functions/getsid.html

Reply:
Hmm actually ok, I think I get it. reading around freedesktop.org suggests that systemd has a concept higher level than Unix Session "Login Session" that holds many processes and process groups and unix sessions. Since each Terminal for e.g. will set itself up as setsid. Wondering if asking LoginServer "hey, what's the session ID for <pid>" might be better in the future, but this should work for now.

Reply:
`sid`s are also used by terminals and other programs. The rational here is that the root `sid` is the login session, and we want to operate on that. This is the approach `LoginServer` took before, I've just move the logic into `SessionManagement`.